serve/mlc_serve/model/tvm_model.py

```diff
@@ -232,7 +238,9 @@ def profile_memory_usage(self, seq_lens):
 
         self.mod["evaluate"](input_ids, positions, seq_lens, self.params)
 
-        return self.get_used_memory()
+        vm_alloc_after = self.get_used_memory()
```

Comment:
sorry there is an incredibly embarrassing hack I did to TVM's `PooledAllocator` based on an assumption that `get_used_memory()` is only called once: 
https://github.com/octoml/tvm/blob/for-mlc-serve-jan12/src/runtime/memory/pooled_allocator.h#L114-L116

This was to change the behavior of `PooledAllocator` before / after memory profiling, to workaround a limitation of the allocator. I have [a branch](https://github.com/octoml/tvm/compare/for-mlc-serve-jan12...rmm-memory-manager?expand=1) that integrates a third-party CUDA allocator https://github.com/rapidsai/rmm to replace `PooledAllocator`.

Reply:
That said, we can make `get_used_memory()` pure by introducing a global function to toggle the allocation mode. I should have done that but I did the shortcut of abusing `get_used_memory()` to signal the end of memory profiling. 

Such global function is also useful even after we land the allocator based on rmm, since tracking statistics for memory allocation is costly and should only be done during memory profiling.

Reply:
Ah, got it.  So, by calling it twice, I'm invalidating the safety net anyways.

I think in that case, I'd lean toward disabling this check altogether, and replacing it with a compile-time tracking of memory usage.  That way, instead of profiling, we could know the maximum size of the live set.

Reply:
I'm curious if such compile-time estimate of memory usage agrees with the "ground truth" that we collect at runtime, given that (I think) the latter is affected by the behavior of the underlying allocator. And my impression of Relax's memory planning has been that it allocates much more than necessary, compared to the allocator in torch.

Reply:
Yeah.  I think that increased memory would show up in the compile-time estimation, because a large part of the footprint is due to operations that aren't done in place.  @slyubomirsky  has been doing a lot of good work for in-place operations, which should give us a path forward on footprint.